File: alphaagent/server/api/data_sync.py
"""Data sync management endpoints."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Body, Query
from fastapi.responses import JSONResponse

from alphaagent.server.core.responses import fail, ok
from alphaagent.server.services import data_sync as service

router = APIRouter(prefix="/data-sync", tags=["data-sync"])


@router.get("/sources")
def sources():
    try:
        return ok(service.list_sources())
    except Exception as exc:
        return _sync_error(exc)


@router.get("/jobs")
def jobs():
    try:
        return ok(service.list_jobs())
    except Exception as exc:
        return _sync_error(exc)


@router.post("/jobs/{job_id}/run")
def run_job(job_id: str, params: dict[str, Any] = Body(default_factory=dict)):
    try:
        return ok(service.run_job(job_id, params))
    except Exception as exc:
        return _sync_error(exc)


@router.post("/batches/run-all")
def run_all(payload: dict[str, Any] = Body(default_factory=dict)):
    try:
        job_ids = payload.get("job_ids")
        return ok(
            service.start_sync_batch(
                profile=str(payload.get("profile") or "core"),
                job_ids=job_ids if isinstance(job_ids, list) and job_ids else None,
                params=payload.get("params") if isinstance(payload.get("params"), dict) else {},
            )
        )
    except Exception as exc:
        return _sync_error(exc)


@router.get("/batches/latest")
def latest_batch():
    try:
        return ok(service.get_latest_sync_batch())
    except Exception as exc:
        return _sync_error(exc)


@router.get("/batches/{batch_id}")
def batch_detail(batch_id: str):
    try:
        return ok(service.get_sync_batch(batch_id))
    except Exception as exc:
        return _sync_error(exc)


@router.post("/jobs/{job_id}/schedule")
def schedule_job(job_id: str, payload: dict[str, Any] = Body(default_factory=dict)):
    try:
        return ok(service.update_job_schedule(job_id, payload.get("schedule_cron")))
    except Exception as exc:
        return _sync_error(exc)


@router.get("/schedules")
def list_schedules():
    try:
        return ok(service.list_schedules())
    except Exception as exc:
        return _sync_error(exc)


@router.post("/schedules")
def create_schedule(payload: dict[str, Any] = Body(default_factory=dict)):
    try:
        return ok(service.create_schedule(payload))
    except Exception as exc:
        return _sync_error(exc)


@router.patch("/schedules/{schedule_id}")
def update_schedule(schedule_id: str, payload: dict[str, Any] = Body(default_factory=dict)):
    try:
        return ok(service.update_schedule(schedule_id, payload))
    except Exception as exc:
        return _sync_error(exc)


@router.delete("/schedules/{schedule_id}")
def delete_schedule(schedule_id: str):
    try:
        return ok(service.delete_schedule(schedule_id))
    except Exception as exc:
        return _sync_error(exc)


@router.post("/schedules/{schedule_id}/run")
def run_schedule(schedule_id: str):
    try:
        return ok(service.run_schedule_now(schedule_id))
    except Exception as exc:
        return _sync_error(exc)


@router.get("/runs")
def runs(limit: int = Query(default=20, ge=1, le=100)):
    try:
        return ok(service.list_runs(limit=limit))
    except Exception as exc:
        return _sync_error(exc)


@router.get("/coverage")
def coverage(force: bool = Query(default=False)):
    try:
        return ok(service.coverage(force_refresh=force))
    except Exception as exc:
        return _sync_error(exc)


@router.get("/usage")
def usage():
    try:
        return ok(service.usage())
    except Exception as exc:
        return _sync_error(exc)


@router.get("/health")
def data_health(force: bool = Query(default=False)):
    try:
        return ok(service.data_health(force_refresh=force))
    except Exception as exc:
        return _sync_error(exc)


def _sync_error(exc: Exception) -> JSONResponse:
    return JSONResponse(
        status_code=503,
        content=fail(
            "DATA_SYNC_UNAVAILABLE",
            "本地数据同步模块暂时不可用，请检查 DATABASE_URL 和 PostgreSQL 连接。",
            {"reason": exc.__class__.__name__},
        ),
    )

File: frontend/src/api/dataSync.ts
import { apiClient, authFetch } from "./client";
import type { SyncCoverage } from "./types";

/** Backend returns plain arrays for these endpoints, not {items, total} wrappers. */
type ListResponse<T> = T[] | { items?: T[] };

function unwrapItems<T>(payload: ListResponse<T>): T[] {
  return Array.isArray(payload) ? payload : payload.items ?? [];
}

export async function fetchSyncSources() {
  return unwrapItems(await apiClient.get<ListResponse<SyncSourceItem>>("/data-sync/sources"));
}

export async function fetchSyncJobs() {
  return unwrapItems(await apiClient.get<ListResponse<SyncJobItem>>("/data-sync/jobs"));
}

export async function fetchSyncRuns(limit = 20) {
  return unwrapItems(await apiClient.get<ListResponse<SyncRunItem>>(`/data-sync/runs?limit=${limit}`));
}

export function fetchSyncCoverage() {
  return apiClient.get<SyncCoverage>("/data-sync/coverage");
}

export function fetchDataUsage() {
  return apiClient.get<DataUsageResponse>("/data-sync/usage");
}

export function fetchDataHealth(force = false) {
  // force=1 绕过服务端 60s 缓存，由「刷新」按钮显式触发
  return apiClient.get<DataHealth>(`/data-sync/health${force ? "?force=1" : ""}`);
}

export function runSyncJob(jobId: string, params: Record<string, unknown> = {}) {
  return apiClient.post<SyncRunItem>(`/data-sync/jobs/${encodeURIComponent(jobId)}/run`, params);
}

export function runAllSyncJobs(
  payload: { profile?: "core" | "all"; job_ids?: string[]; params?: Record<string, unknown> } = {},
) {
  return apiClient.post<SyncBatchStatus>("/data-sync/batches/run-all", payload);
}

// ── Batch schedules (unified incremental sync slots) ──

export function fetchSyncSchedules() {
  return apiClient.get<BatchSchedule[]>("/data-sync/schedules");
}

export function createSyncSchedule(payload: Partial<BatchSchedule>) {
  return apiClient.post<BatchSchedule>("/data-sync/schedules", payload);
}

export function updateSyncSchedule(id: string, payload: Partial<BatchSchedule>) {
  return apiClient.patch<BatchSchedule>(`/data-sync/schedules/${encodeURIComponent(id)}`, payload);
}

export function deleteSyncSchedule(id: string) {
  return apiClient.del<{ id: string }>(`/data-sync/schedules/${encodeURIComponent(id)}`);
}

export function runSyncSchedule(id: string) {
  return apiClient.post<SyncBatchStatus>(`/data-sync/schedules/${encodeURIComponent(id)}/run`);
}

export async function fetchLatestSyncBatch() {
  const response = await authFetch("/data-sync/batches/latest");
  if (!response.ok) {
    throw new Error(`请求失败: ${response.status} ${response.statusText}`);
  }
  const body = await response.json();
  if (!body.success) {
    throw new Error(body.error?.message ?? "同步批次状态读取失败");
  }
  return (body.data ?? null) as SyncBatchStatus | null;
}

export function fetchSyncBatch(batchId: string) {
  return apiClient.get<SyncBatchStatus>(`/data-sync/batches/${encodeURIComponent(batchId)}`);
}

// ── Types matching actual backend responses ──

export interface SyncSourceItem {
  id: string;
  name: string;
  kind: string;
  base_url?: string;
  enabled: boolean;
  priority: number;
  status: string;
  message?: string;
  checked_at?: string;
}

export interface SyncJobItem {
  id: string;
  name: string;
  description: string;
  source_id: string;
  target_table: string;
  enabled: boolean;
  default_params: Record<string, unknown>;
  schedule_cron?: string | null;
  last_status?: string | null;
  last_run_id?: string | null;
  last_started_at?: string | null;
  message?: string;
}

export interface SyncRunItem {
  id?: string | number;
  run_id?: string | number;
  job_id: string;
  status: string;
  params?: Record<string, unknown>;
  started_at: string;
  finished_at?: string | null;
  rows_read: number;
  rows_written: number;
  message?: string | null;
  error?: string | null;
  error_type?: string | null;
}

export interface DataUsageCapability {
  name: string;
  table: string;
  description: string;
  status: string;
  count: number;
  message?: string;
}

export interface DataUsageResponse {
  capabilities: DataUsageCapability[];
  coverage: Record<string, { count: number; last_updated?: string }>;
}

export interface SyncBatchJobStatus {
  job_id: string;
  status: "pending" | "running" | "succeeded" | "failed" | string;
  started_at?: string | null;
  finished_at?: string | null;
  rows_read: number;
  rows_written: number;
  progress_current?: number;
  progress_total?: number;
  progress_pct?: number;
  stage?: string;
  current_label?: string;
  sample_items?: SyncProgressSample[];
  message?: string;
  run_id?: string | number | null;
  error_type?: string;
}

export type SyncProgressSample = Record<string, string | number | boolean | string[] | null | undefined>;

export interface SyncBatchStatus {
  id: string;
  profile: "core" | "all" | string;
  source?: string;
  schedule_id?: string | null;
  concurrency?: number;
  status: "running" | "succeeded" | "failed" | string;
  created_at: string;
  started_at?: string | null;
  finished_at?: string | null;
  current_job_id?: string | null;
  total_jobs: number;
  completed_jobs: number;
  succeeded_jobs: number;
  failed_jobs: number;
  rows_read: number;
  rows_written: number;
  progress_pct: number;
  message?: string;
  jobs: SyncBatchJobStatus[];
}

/** A unified incremental sync slot (replaces per-job crons). */
export interface BatchSchedule {
  id: string;
  name: string;
  cron: string;
  action?: "sync" | "low_suction_live_scan" | string;
  job_ids: string[];
  enabled: boolean;
  concurrency: number;
  last_status?: string | null;
  last_started_at?: string | null;
  last_finished_at?: string | null;
  last_message?: string | null;
}

export type DataHealthSeverity = "fresh" | "stale" | "empty" | "future" | "unknown" | string;
export type DataHealthLevel = "green" | "yellow" | "red" | string;

export interface DataHealthJob {
  job_id: string;
  name: string;
  cadence: string;
  category: string;
  is_stale: boolean;
  severity: DataHealthSeverity;
  local_latest?: string | null;
  staleness_days: number;
  reason: string;
  recommended: boolean;
}

export interface DataHealthCategory {
  key: string;
  label: string;
  health: DataHealthLevel;
  jobs: DataHealthJob[];
}

export interface DataHealth {
  generated_at: string;
  overall: {
    health: DataHealthLevel;
    summary: string;
    is_empty_database: boolean;
    empty_core_tables: string[];
    stale_count: number;
    fresh_count: number;
    recommended_count: number;
  };
  market_context: {
    now: string;
    latest_trade_date?: string | null;
    latest_daily_trade_date?: string | null;
    latest_complete_trade_date?: string | null;
    latest_trade_date_symbol_count?: number | null;
    min_complete_daily_symbol_count?: number;
    is_disclosure_season: boolean;
    trade_calendar_source: string;
  };
  categories: DataHealthCategory[];
  recommended: {
    job_ids: string[];
    count: number;
    rationale: string;
  };
  bootstrap: {
    needed: boolean;
    core_profile_job_ids: string[];
    message: string | null;
  };
}
